prov/rxm/src/rxm_cq.c

```diff
@@ -1169,27 +1187,68 @@ static ssize_t rxm_atomic_send_resp(struct rxm_ep *rxm_ep,
 	return ret;
 }
 
-static void rxm_do_atomic(struct rxm_pkt *pkt, void *dst, void *src,
-			  void *cmp, void *res, size_t count,
-			  enum fi_datatype datatype, enum fi_op op)
+static int rxm_do_atomic(uint8_t op, void *dst, void *src, void *cmp, void *res,
+			 size_t count, enum fi_datatype datatype,
+			 enum fi_op amo_op)
 {
-	switch (pkt->hdr.op) {
+	switch (op) {
 	case ofi_op_atomic:
-		assert(ofi_atomic_iswrite_op(op));
-		ofi_atomic_write_handler(op, datatype, dst, src, count);
+		assert(ofi_atomic_iswrite_op(amo_op));
+		ofi_atomic_write_handler(amo_op, datatype, dst, src, count);
 		break;
 	case ofi_op_atomic_fetch:
-		assert(ofi_atomic_isreadwrite_op(op));
-		ofi_atomic_readwrite_handler(op, datatype, dst, src, res, count);
+		assert(ofi_atomic_isreadwrite_op(amo_op));
+		ofi_atomic_readwrite_handler(amo_op, datatype, dst, src, res,
+					     count);
 		break;
 	case ofi_op_atomic_compare:
-		assert(ofi_atomic_isswap_op(op));
-		ofi_atomic_swap_handler(op, datatype, dst, src, cmp, res, count);
+		assert(ofi_atomic_isswap_op(amo_op));
+		ofi_atomic_swap_handler(amo_op, datatype, dst, src, cmp, res,
+					count);
 		break;
 	default:
 		/* Validated prior to calling function */
 		break;
 	}
+
+	return FI_SUCCESS;
+}
+
+static int rxm_do_device_mem_atomic(struct rxm_mr *dev_mr, uint8_t op,
+				    void *dev_dst, void *src, void *cmp,
+				    void *res, size_t amo_count,
+				    enum fi_datatype datatype,
+				    enum fi_op amo_op, size_t amo_op_size)
+{
+	void *bounce_buf;
+	ssize_t ret __attribute__((unused));
+	struct iovec iov = {
+		.iov_base = dev_dst,
+		.iov_len = amo_op_size,
+	};
+
+	bounce_buf = malloc(amo_op_size);
+	if (!bounce_buf)
+		return -FI_ENOMEM;
+
+	fastlock_acquire(&dev_mr->amo_lock);
+	ret = ofi_copy_from_hmem_iov(bounce_buf, amo_op_size, dev_mr->iface, 0,
+				    &iov, 1, 0);
+	assert(ret == amo_op_size);
+
+	ret = rxm_do_atomic(op, bounce_buf, src, cmp, res, amo_count, datatype,
+			    amo_op);
+	assert(ret == FI_SUCCESS);
+
+	ret = ofi_copy_to_hmem_iov(dev_mr->iface, 0, &iov, 1, 0, bounce_buf,
+				   amo_op_size);
+
+	fastlock_release(&dev_mr->amo_lock);
```

Comment:
Having these locks does not make the operation atomic wrt whatever may be happening on the device.  I believe the CPU access is already serialized coming through this call, so the lock shouldn't be needed.